project/data_analysis/python/so/so_prepare_sim_data.py

```diff
@@ -0,0 +1,347 @@
+
+The code makes use of the SO noise calculator: so_noise_calculator_public_20180822.
+"""
+from pspy import pspy_utils, so_dict, so_spectra, so_map, so_window
+from itertools import combinations_with_replacement as cwr
+import so_noise_calculator_public_20180822 as noise_calc
+from pspipe_utils import pspipe_list, best_fits
+import matplotlib.pyplot as plt
+from copy import deepcopy
+import numpy as np
+import camb
+import sys
+
+d = so_dict.so_dict()
+d.read_from_file(sys.argv[1])
+
+type = d["type"]
+surveys = d["surveys"]
+
+# Define and create the directories to save inputs for the simulations
+windows_dir = d["windows_dir"]
+bestfit_dir = "best_fits"
+noise_model_dir = "noise_model"
+plot_dir = "plots"
+beam_dir = d["beam_dir"]
+binning_dir = d["binning_dir"]
+passband_dir = d["passband_dir"]
+
+pspy_utils.create_directory(windows_dir)
+pspy_utils.create_directory(bestfit_dir)
+pspy_utils.create_directory(noise_model_dir)
+pspy_utils.create_directory(beam_dir)
+pspy_utils.create_directory(binning_dir)
+pspy_utils.create_directory(passband_dir)
+
+# Linestyles used to make the difference between LAT and Planck in the plots
+linestyle = {
+    "LAT": "solid",
+    "Planck": "dashed"
+}
+
+# Get survey, array lists & spectra name list
+n_arrays, sv_list, ar_list = pspipe_list.get_arrays_list(d)
+spec_name_list = pspipe_list.get_spec_name_list(d, char = "_")
+
+spectra = ["TT", "TE", "TB", "ET", "BT", "EE", "EB", "BE", "BB"]
+
+# We start with SO, we have to specify a sensitivity mode (2: goal, 1: baseline), and f_sky
+# both parameters are specified in the dictionnary
+sensitivity_mode = d["sensitivity_mode"]
+f_sky_LAT = d["f_sky_LAT"]
+
+ell_min, ell_max = 2, d["lmax"] + 500
+delta_ell = 1
+# We use the SO noise calculator to compute the expected noise in temperature and polarisation.
+ell, n_ell_t_LAT, n_ell_pol_LAT, map_wn = noise_calc.Simons_Observatory_V3_LA_noise(sensitivity_mode,
+                                                                                    f_sky_LAT,
+                                                                                    ell_min,
+                                                                                    ell_max,
+                                                                                    delta_ell,
+                                                                                    N_LF=1.,
+                                                                                    N_MF=4.,
+                                                                                    N_UHF=2.,
+                                                                                    apply_beam_correction=False,
+                                                                                    apply_kludge_correction=True)
+
+# Initialize the dictionnaries with zeros
+n_ell_t = {spec_name: ell * 0 for spec_name in spec_name_list}
```

Comment:
here I will make sure `n_ell_t` is a float array by either writing `ell * 0.0` (I have no clue about the data type of `ell` returned by `so_noise_calculator`) or `np.zeros_like(ell, dtype=float)`